fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.lucene.index.ConcurrentMergeScheduler;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.MergePolicy;
+import org.apache.lucene.index.TieredMergePolicy;
+import org.apache.lucene.util.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static com.apple.foundationdb.record.lucene.DirectoryCommitCheckAsync.getOrCreateDirectoryCommitCheckAsync;
+
+/**
+ * This class closes the writer after commit.  The goal here is to only flush the results
+ * at the end of the commit.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class IndexWriterCommitCheckAsync implements FDBRecordContext.CommitCheckAsync {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneIndexMaintainer.class);
+    protected final IndexWriter indexWriter;
+    protected final Executor executor;
+
+
+    /**
+     * Creates an index writer config with merge configurations that limit the amount of data in a segment.
+     *
+     * @param analyzer analyzer
+     * @param directoryCommitCheckAsync directoryCommitCheckAsync
+     * @param executor executor
+     * @throws IOException exception
+     */
+    public IndexWriterCommitCheckAsync(@Nonnull Analyzer analyzer, @Nonnull DirectoryCommitCheckAsync directoryCommitCheckAsync, Executor executor) throws IOException {
+        TieredMergePolicy tieredMergePolicy = new TieredMergePolicy();
+        tieredMergePolicy.setMaxMergedSegmentMB(5.00);
+        tieredMergePolicy.setMaxMergeAtOnceExplicit(2);
+        tieredMergePolicy.setNoCFSRatio(1.00);
+        IndexWriterConfig indexWriterConfig = new IndexWriterConfig(analyzer);
+        indexWriterConfig.setUseCompoundFile(true);
+        indexWriterConfig.setMergePolicy(tieredMergePolicy);
+        indexWriterConfig.setMergeScheduler(new ConcurrentMergeScheduler() {
+            @Override
+            protected void doMerge(final IndexWriter writer, final MergePolicy.OneMerge merge) throws IOException {
+                merge.segments.forEach( (segmentCommitInfo) -> LOGGER.trace("segmentInfo={}", segmentCommitInfo.info.name));
+                super.doMerge(writer, merge);
+            }
+        });
+        indexWriterConfig.setCodec(new Lucene70Codec(Lucene50StoredFieldsFormat.Mode.BEST_COMPRESSION));
+        this.indexWriter = new IndexWriter(directoryCommitCheckAsync.getDirectory(), indexWriterConfig);
+        this.executor = executor;
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<Void> checkAsync() {
+        LOGGER.trace("closing writer check");
+        return indexWriter.isOpen() ?
+               CompletableFuture.runAsync(() -> IOUtils.closeWhileHandlingException(indexWriter), executor) :
```

Comment:
Does `IndexWriter.close` need to happen on transaction rollback (that is, `close` without `commit`)?

That was the idea behind having the session objects be something narrower than `Object` so they could have a lifecycle. Maybe the Java GC does a good enough job?

Reply:
It should be collected by GC.  

`    // Ensure that only one thread actually gets to do the
    // closing
    if (shouldClose(true)) {
      try {
        if (infoStream.isEnabled("IW")) {
          infoStream.message("IW", "now flush at close");
        }

        flush(true, true);
        waitForMerges();
        commitInternal(config.getMergePolicy());
        rollbackInternal(); // ie close, since we just committed
      } catch (Throwable t) {
        // Be certain to close the index on any exception
        try {
          rollbackInternal();
        } catch (Throwable t1) {
          t.addSuppressed(t1);
        }
        throw t;
      }
    }`